### PR TITLE
Adds a `cors` decorator to manage cross-origin requests

### PR DESCRIPTION
**Pull request by shreyas-satish**
Usage:

```python
from flask import Response
import app

def check_origin(origin):
    # check if origin is allowed access in client app
    return True

@app.route('/')
@cors(check_origin)
def index():
    return Response()
````

### CODE REVIEW

File: coaster/views.py [hunk 1]
```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
                 return result
         return decorated_function
     return inner
+
+def cors(is_origin_allowed=lambda origin: False,
+    allowed_methods='HEAD, OPTIONS, GET, POST, PUT, PATCH, DELETE',
```
Comment (jace) [inline coaster/views.py]: Why isn't `allowed_methods` a list?

File: coaster/views.py [hunk 2]
```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
                 return result
         return decorated_function
     return inner
+
+def cors(is_origin_allowed=lambda origin: False,
```
Comment (jace) [inline coaster/views.py]: Make the parameter mandatory. If it defaults to False, making the method uncallable, there's no reason to have the default.

File: coaster/views.py [hunk 3]
```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
...
+    allowed_methods='HEAD, OPTIONS, GET, POST, PUT, PATCH, DELETE',
+    allowed_headers='Accept, Accept-Language, Content-Language, Content-Type, X-Requested-With',
```
Comment (jace) [inline coaster/views.py]: Should also be a list.

File: coaster/views.py [hunk 4]
```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
...
+        def wrapper(*args, **kwargs):
+            origin = request.headers.get('Origin')
+            if request.method not in allowed_methods or not is_origin_allowed(origin):
+                abort(401)
+
+            if request.method == 'OPTIONS':
+                # pre-flight request
+                resp = Response()
+            else:
+                resp = f(*args, **kwargs)
```
Comment (jace) [inline coaster/views.py]: The view may not return a response object. Flask allows views to return any of the following:

1. A response object
2. A string (response body)
3. A 2-tuple consisting of a string and a status code
4. A 3-tuple consisting of a string, status code and HTTP headers

See `coaster.views.render_with` for how we handle these variations.

File: coaster/views.py [hunk 5]
```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
...
+
+            resp.headers['Access-Control-Allow-Origin'] = origin
+            resp.headers['Access-Control-Allow-Methods'] = allowed_methods
+            resp.headers['Access-Control-Allow-Headers'] = allowed_headers
```
Comment (jace) [inline coaster/views.py]: Use `', '.join(allowed_…)`, assuming those are lists now.

File: coaster/views.py [hunk 6]
```diff
@@ -574,3 +574,41 @@ def decorated_function(*args, **kwargs):
...
+    max_age=None):
+    """
+    Adds CORS headers to the decorated view function.
+
+    :param is_origin_allowed: A handler function that receives the origin as a parameter and
+    is expected to return a boolean value to assert if the given origin has access to the
+    requested resource
+    :param allowed_methods: A string of comma-separated HTTP methods that are allowed for this origin
+    :param allowed_headers: A string of comma-separated HTTP headers that are allowed for this origin
+    :param max_age: Maximum number of seconds the result for the pre-flight request can be cached
```
Comment (jace) [inline coaster/views.py]: Change the parameter names to `validator`, `methods` and `headers` (`max_age` stays).

File: coaster/views.py (hunk 4 above)
Comment (shreyas-satish) [inline coaster/views.py]: Using `make_response` in case the view's return value is not a response object.

File: coaster/views.py (hunk 5 above)
Comment (shreyas-satish) [inline coaster/views.py]: Fixed.

File: coaster/views.py (hunk 6 above)
Comment (shreyas-satish) [inline coaster/views.py]: Fixed.

File: coaster/views.py (hunk 3 above)
Comment (shreyas-satish) [inline coaster/views.py]: Fixed.

File: coaster/views.py (hunk 2 above)
Comment (shreyas-satish) [inline coaster/views.py]: Fixed.

**Comment by shreyas-satish** [comment]
Also, added `Origin` in the `Vary` header as per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin.

**Pull request merged**